File: src/shared/types/message.ts
/**
 * Message domain types.
 * Represents chat messages and related data.
 */

import type { Attachment } from "./attachment";
import type { BaseEntity, ISODateString, MessageStatus, UUID } from "./common";
import type { User } from "./user";

/** Reaction to a message */
export interface Reaction {
  readonly createdAt: ISODateString;
  readonly emoji: string;
  readonly id: UUID;
  readonly messageId: UUID;
  readonly userId: UUID;
  readonly username: string;
}

/** Mention in a message */
export interface Mention {
  readonly id: UUID;
  readonly messageId: UUID;
  readonly userId: UUID;
  readonly username: string;
}

/** Read receipt for a message */
export interface ReadReceipt {
  readonly id: UUID;
  readonly messageId: UUID;
  readonly readAt: ISODateString;
  readonly userId: UUID;
}

/** Base message entity */
export interface Message extends BaseEntity {
  readonly content: string | null;
  readonly isDeleted: boolean;
  readonly isEdited: boolean;
  readonly replyToId: UUID | null;
  readonly senderId: UUID;
  readonly status: MessageStatus;
}

/** Message with all related data for display */
export interface MessageWithDetails extends Message {
  readonly attachments: readonly Attachment[];
  readonly mentions: readonly Mention[];
  readonly reactions: readonly Reaction[];
  readonly readBy: readonly ReadReceipt[];
  readonly replyTo: Message | null;
  readonly sender: User;
}

/** Data for creating a new message */
export interface CreateMessageInput {
  readonly attachmentIds?: readonly UUID[];
  readonly content: string;
  readonly mentionedUserIds?: readonly UUID[];
  readonly replyToId?: UUID;
}

/** Data for updating a message */
export interface UpdateMessageInput {
  readonly content: string;
}

/** Message search parameters */
export interface MessageSearchParams {
  readonly endDate?: ISODateString;
  readonly hasAttachments?: boolean;
  readonly limit?: number;
  readonly offset?: number;
  readonly query: string;
  readonly senderId?: UUID;
  readonly startDate?: ISODateString;
}

/** Grouped messages by date */
export interface MessageGroup {
  readonly date: string;
  readonly messages: readonly MessageWithDetails[];
}

/** Message event for real-time updates */
export type MessageEvent =
  | { readonly type: "created"; readonly message: MessageWithDetails }
  | { readonly type: "updated"; readonly message: MessageWithDetails }
  | { readonly type: "deleted"; readonly messageId: UUID }
  | { readonly type: "reaction_added"; readonly reaction: Reaction }
  | { readonly type: "reaction_removed"; readonly reactionId: UUID };

File: src-tauri/src/models/message.rs
//! Message domain model.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use super::{Attachment, User};

/// Message delivery status.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MessageStatus {
    Sending,
    Sent,
    Delivered,
    Read,
    Failed,
}

impl Default for MessageStatus {
    fn default() -> Self {
        Self::Sending
    }
}

/// Base message entity.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Message {
    pub id: Uuid,
    pub sender_id: Uuid,
    pub content: Option<String>,
    pub reply_to_id: Option<Uuid>,
    pub is_edited: bool,
    pub is_deleted: bool,
    pub status: MessageStatus,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl Message {
    /// Create a new message.
    pub fn new(sender_id: Uuid, content: String, reply_to_id: Option<Uuid>) -> Self {
        let now = Utc::now();
        Self {
            id: Uuid::new_v4(),
            sender_id,
            content: Some(content),
            reply_to_id,
            is_edited: false,
            is_deleted: false,
            status: MessageStatus::Sending,
            created_at: now,
            updated_at: now,
        }
    }
}

/// Message with all related data for display.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageWithDetails {
    #[serde(flatten)]
    pub message: Message,
    pub sender: User,
    pub reply_to: Option<Box<Message>>,
    pub attachments: Vec<Attachment>,
    pub reactions: Vec<Reaction>,
    pub mentions: Vec<Mention>,
    pub read_by: Vec<ReadReceipt>,
}

/// Reaction to a message.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Reaction {
    pub id: Uuid,
    pub message_id: Uuid,
    pub user_id: Uuid,
    pub username: String,
    pub emoji: String,
    pub created_at: DateTime<Utc>,
}

impl Reaction {
    /// Create a new reaction.
    pub fn new(message_id: Uuid, user_id: Uuid, username: String, emoji: String) -> Self {
        Self {
            id: Uuid::new_v4(),
            message_id,
            user_id,
            username,
            emoji,
            created_at: Utc::now(),
        }
    }
}

/// Mention in a message.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Mention {
    pub id: Uuid,
    pub message_id: Uuid,
    pub user_id: Uuid,
    pub username: String,
}

/// Read receipt for a message.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReadReceipt {
    pub id: Uuid,
    pub message_id: Uuid,
    pub user_id: Uuid,
    pub read_at: DateTime<Utc>,
}

/// Input for creating a new message.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateMessageInput {
    pub content: String,
    pub reply_to_id: Option<Uuid>,
    pub mentioned_user_ids: Option<Vec<Uuid>>,
    pub attachment_ids: Option<Vec<Uuid>>,
}

/// Input for updating a message.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateMessageInput {
    pub content: String,
}

/// Message search parameters.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageSearchParams {
    pub query: String,
    pub sender_id: Option<Uuid>,
    pub start_date: Option<DateTime<Utc>>,
    pub end_date: Option<DateTime<Utc>>,
    pub has_attachments: Option<bool>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

/// Paginated response wrapper.
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PaginatedMessages {
    pub items: Vec<MessageWithDetails>,
    pub total: i64,
    pub has_more: bool,
}
